src/lib/seam/connect/seam-http-request.ts

```diff
@@ -0,0 +1,127 @@
+import { serializeUrlSearchParams } from '@seamapi/url-search-params-serializer'
+import type { Method } from 'axios'
+
+import type { Client } from './client.js'
+import type { SeamHttpRequestOptions } from './options.js'
+import { resolveActionAttempt } from './resolve-action-attempt.js'
+import { SeamHttpActionAttempts } from './routes/index.js'
+
+interface SeamHttpRequestParent {
+  readonly client: Client
+  readonly defaults: Required<SeamHttpRequestOptions>
+}
+
+export interface SeamHttpRequestConfig<TBody, TResponseKey> {
+  readonly path: string
+  readonly method: Method
+  readonly body?: TBody
+  readonly query?: undefined | Record<string, unknown>
+  readonly responseKey: TResponseKey
+  readonly options?: Pick<SeamHttpRequestOptions, 'waitForActionAttempt'>
+}
+
+export class SeamHttpRequest<
+  const TBody,
+  const TResponse,
+  const TResponseKey extends keyof TResponse | undefined,
+> implements
+    PromiseLike<
+      TResponseKey extends keyof TResponse ? TResponse[TResponseKey] : undefined
+    >
+{
+  readonly #parent: SeamHttpRequestParent
+  readonly #config: SeamHttpRequestConfig<TBody, TResponseKey>
+
+  constructor(
+    parent: SeamHttpRequestParent,
+    config: SeamHttpRequestConfig<TBody, TResponseKey>,
+  ) {
+    this.#parent = parent
+    this.#config = config
+  }
+
+  public get responseKey(): TResponseKey {
+    return this.#config.responseKey
+  }
+
+  public get url(): URL {
+    const { client } = this.#parent
+    const baseUrl = client.defaults.baseURL
+    if (baseUrl === undefined) {
+      throw new Error('baseUrl is required')
+    }
+
+    const query = this.#config.query
+    if (query === undefined) {
+      return new URL(this.#config.path, baseUrl)
+    }
+    return new URL(
+      `${this.#config.path}?${serializeUrlSearchParams(query)}`,
+      baseUrl,
+    )
+  }
+
+  public get method(): Method {
+    return this.#config.method
+  }
+
+  public get body(): TBody {
+    return this.#config.body as TBody
```

Comment:
Just trying to understand the advanced type stuff we are adding here. Why is this cast needed if it's already part of the generic type?

Reply:
`body` is an optional property on `SeamHttpRequestConfig`, so typescript incorrectly thinks that `this.#config.body` is `TBody | undefined` instead of `TBody` because it cannot know that `TBody` itself is set to `SomeParams | undefined` in this scenario.

Reply:
> typescript incorrectly thinks that this.#config.body is TBody | undefined instead of TBody

Can't `seamHttpReq.body` be undefined, for example for get requests without a body?